commons/src/main/java/org/infinispan/commons/api/functional/Listeners.java

```diff
@@ -0,0 +1,221 @@
+package org.infinispan.commons.api.functional;
+
+import org.infinispan.commons.api.functional.EntryView.ReadEntryView;
+
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+/**
+ * Listener definitions.
+ *
+ * DESIGN RATIONALE:
+ * <ul>
+ *    <li>The current set of listener events that can be fired are related
+ *    to modifications, and deciding between create or modify, or providing
+ *    removed entry information, require reading the previous value. Hence,
+ *    the current set of listeners can only be provided for read-write maps.
+ *    In the future, if for example cache entry visited events are to be
+ *    supported, those would need to be associated with either a read or
+ *    read-write map, and hence the listeners interface would most likely
+ *    be split up.
```

Comment:
> Activation, passivation, and expiration events would also have to be available for read-only maps. 

Activation only makes sense for read-only and read-write maps, since activation is the result of reading something. To allow activation events for a write-only view of the map would be a user error so should be not allowed. 

Passivation makes sense for all read-only, read-write and write-only since entries can be passivated as a result of both a write and read operation.

Expiration events would apply to all too. Even write because if you set a lifespan of 0, a write-only operation would cause an expiration event.

Support for these events are TODO.

> Actually, the fact that one object only has a read-only cache view doesn't mean that other objects can't modify the cache, so I'd argue that all listeners should be available for all views.

No, not all listeners should be available to all. For example, if you only call write-only operations, you should not be allowed to have a modify event, because a modification or an update can only be deduced if you read the previous value. Hence, to allow modify events to be listened on a write-only map would a user error. The design proposed here avoids this kind of errors.

@wburns I'll reply to your comment a bit later on...